ovsdb/client.go

```diff
@@ -146,7 +168,15 @@ func (c *Client) rpc(ctx context.Context, method string, out interface{}, args .
 	// Await RPC completion or cancelation.
 	select {
 	case <-ctx.Done():
-		// RPC canceled.  Producer cleans up the callback.
+		// RPC canceled.  Clean up the callback in case no message ever arrives
+		// with its request ID, so we don't leak callbacks.  If the other case
+		// in the select fires (meaning we got a matching RPC response), it's
+		// the producer's responsibility to clean up the callback.
```

Comment:
Can we do that cleanup in Producer side always, just to make in consistent both success and failed cases?  If not, what is the challenge to do so?

Reply:
The problem is that if no message ever arrives for a given request ID, the producer will never trigger its callback for that response ID, and thus, never clean up the callback hook.

I think having the consumer clean up when it no longer wants to hear about the response is probably the most sane approach, even if it's a little awkward.

Reply:
Makes sense.  Then, how about moving the callback cancelation logic to the consumer side all together, meaning let consumer cancel the callback even when the success case.  Let me know your thoughts on that.  I don't think it's usual to drag this PR too long but just want to sync up with you on that point.

Reply:
Yeah, I was thinking of having a `defer cleanupCallback()` right after the `addCallback` call, and then getting rid of the cleanup in `doCallback`. That takes care of the leaks (including the one if `Send` fails) but you're right that it introduces a potential race condition. I guess the thing to do in that case is to leave the `delete` call in `doCallback` intact -- in some cases it gets called twice, but you'll neither leak it nor can it be called in a race between returning from `doCallback` and the cleanup on the consumer.